### DISCUSSION TITLE
How to count learned model parameters

### DISCUSSION THREAD

**Discussion by addisonklinke** [onnx/onnx#2985] [Q&A]
# Ask a Question

### Question
I would like to count to total number of learned (i.e. trainable) parameters in a model given its ONNX file

### Further information
- Relevant Area: model usage, best practices, operators

- Is this issue related to a specific model? No, this is a generic question about ONNX graphs

### Notes

In Pytorch you can do

```
model = Net()
total_params = sum(p for p in model.parameters())
learnable_params = sum(p for p in model.parameters() if p.requires_grad)
```

Since the `.onnx` file of a model includes its weights, it seems that there should be a similar method to analyze a loaded model. I understand that the `requires_grad` attribute would no longer be present since a loaded ONNX model is not associated with a training. However, if possible, it would still be useful to distinguish between total parameters (that contribute to FLOPs) and learnable parameters (that were updated via backprop during training)

```
import onnx

model = onnx.load('/path/to/my_model.onnx')
total_params = ...
learnable_params = ...
```

Are there an existing class methods or best practices to fill in the `...` sections above? I am open to non-Python approaches

**Comment by askhade** [comment]
It should be easy to add a utility for this. I think we can utilize the Differentiable tag to count the learnable parameters. You can learn more about this tag here: https://github.com/onnx/onnx/blob/master/docs/DefineDifferentiability.md
@wschin for comments.

**Reply by wschin** [reply]
ONNX Training Spec introduces [TrainingInfoProto](https://github.com/onnx/onnx/blob/33fd678c6c38baa3e2d9fd5addf4c56cd681a87a/onnx/onnx-ml.proto#L232) with a new field called [update binding](https://github.com/onnx/onnx/blob/33fd678c6c38baa3e2d9fd5addf4c56cd681a87a/onnx/onnx-ml.proto#L328). A binding is simply a key-value string pair and it represents an `assignment` from `value` to `key` (i.e., `key_tensor = value_tensor`). For a model file, it contains [multiple TrainingInfoProto](https://github.com/onnx/onnx/blob/33fd678c6c38baa3e2d9fd5addf4c56cd681a87a/onnx/onnx-ml.proto#L391). If an initializer's name can be found as key in any `update_binding` in those TrainingInfoProto, then that initializer is trainable.

On the other hand, like @askhade said, we can use differentiable tags to search for tensors which can have gradient (the ability of having gradient is not 100% equivalent to being trainable). If an initializer is a differentiable input of an operator, then that initializer can have gradient.
